### ISSUE TITLE
Add "load last open tabs" feature

### ISSUE THREAD

**Issue by andronov-alexey** [CoatiSoftware/Sourcetrail#851]
With behavior analogous to browsers
Is everybody ok if I implement this feature?

**Comment by egraether** [comment]
Yes, that's ok. By analogous to browsers you mean that the user is prompted "Should recently open tabs be reopened?" after opening the project?

`Application::handleMessage(MessageLoadProject*)` is probably a good place to start and then dispatching `MessageTabOpenWith(Id tokenId)` for each tab. However a `tokenId` might change when the project is reindexed, so it's better to store tab information persistently using names. This can be done with, `StorageAccess::getNameHierarchyForNodeId()` and `StorageAccess::getNodeIdForNameHierarchy()`.

The hard question is where to persistently store this information. I can think of a couple places:
* `project.srctrlprj`: XML file. Easy to store, but maybe a bad place because some users may have it under version control.
*  `project.srctrldb`: Sqlite file. This is where the project index is stored. Probably a bad place since we create `.srctrldb_tmp` files during indexing and then switch files when reindexing.
* `project.srctrlbm`: Sqlite file. Contains the bookmarks. Would be a good fit, since it also contains per project user data. But this file was probably a bit over-engineered and it's cumbersome to work with SQL for such simple data. 
* `ApplicationSettings.xml`: XML file. Contains the user preferences and also `recent projects`. Would be a good fit, because it's simple to add new items. 
* Or using a new file that is either stored next to the project or within the user data directory (were `ApplicationSettings.xml` is stored).

I think it's probably easiest to store them either within `ApplicationSettings` or in a file that is in the same directory as `ApplicationSettings.xml`. Any opinions? @mlangkabel?

**Comment by mlangkabel** [comment]
I think `ApplicationSettings.xml` would be the best place to store this info. My suggestion:
Give every project an ID on creation (or when a project that doesn't have an id yet is loaded (possible via fallback value when reading that value from the project settings). e.g.
```
<?xml version="1.0" encoding="utf-8" ?>
<config>
    <id>fe6fdec0-1f2e-4f47-974c-65beb07f797a</id>
    <source_groups>
        ...
    </source_groups>
    <version>8</version>
</config>
```
To generate random unique IDs you can use the same mechanism that is used to generate the IDs of source groups in the project settings.

And in the application settings:
```
<?xml version="1.0" encoding="utf-8" ?>
<config>
    <projects>
        <project_fe6fdec0-1f2e-4f47-974c-65beb07f797a>
            <open_tabs>
                <open_tab>::\tmio\ts\tp\tnnumberIn\tsint\tp()</open_tab>
            <open_tabs>
        </project_fe6fdec0-1f2e-4f47-974c-65beb07f797a>
    </projects>
    <application>
        ...   
```
The stange string in the `open_tab` element is the serialized name hierarchy of the last active symbol for that tab.

**Comment by andronov-alexey** [comment]
> By analogous to browsers you mean that the user is prompted "Should recently open tabs be reopened?" after opening the project

No, just a checkbox in settings "Restore previous session" which is off by default

Thanks for the tips, they are of a great help

**Comment by Alvoda** [comment]
Hello Alexey,

are you still working on this Issue? If not i would try to implement this feature.

**Comment by andronov-alexey** [comment]
Yeah, I'm working on it :)

**Comment by andronov-alexey** [comment]
Hey, guys @egraether @mlangkabel 
I've been investigating the task and here's what I have to the present day:
### 1) Original suggested solution
Storing tabs nodes as suggested is works but with a  big drawback - storing each tabs' `SearchMatch`'es lacks a lot of tab state information namely: current view state of graph and code views (influenced by `Command::ORDER_ADAPT` and `Command::ORDER_VIEW` messages), custom trails and special views ("Overview", "Legend", etc.)

I believe I've found a better solution:
### 2) Store last `Command::ORDER_ACTIVATE` messages with all subsequent `Command::ORDER_ADAPT` and `Command::ORDER_VIEW` messages
pros: we can restore full previous state for all tabs
cons: we need to implement serialization for all children of `MessageBase` which induces
	a) adding the need for non-trivial and fragile versioning of serialized data (every added/deleted field in `MessageBase` and its derivatives affects the serialization)
	b) cluttering "ApplicationSettings.xml" with a lot of `MessageBase` implementation details

So versioning would be a lot of effort not worthy of a simple goal of storing/loading tabs

I also considered:
### 3) Saving and restoring graph and code components but they too will need large amount of complicated and fragile code

So it seems the second solution would be the best for user but too complicated in implementation
My question is what do you think would be the best way to solve the original feature given the information above?
Maybe I'm missing something or made wrong assumptions?
Thanks for any help

**Comment by mlangkabel** [comment]
Hm, you are making a good point. Having to keep track of message versions does not sound that great. @egraether, what do you think? Do we need it like this? Is there another way?

**Comment by egraether** [comment]
Sorry for the late response. I somehow missed/forgot this.

I think we should stick to the original suggestion for now and only restore the last active symbol per tab based on the name of the current active symbol.

If desired we can later also store the state information by serializing the messages stored per tab in the `UndoRedoController`. This can be fully implemented within the `UndoRedoController` without touching the implementation of the message types (e.g. `std::wstring UndoRedoController::serializeCommands() const`).
